src/clients/node/src/node.zig

```diff
@@ -15,783 +19,472 @@ const CreateTransfersResult = tb.CreateTransfersResult;
-    if (c.napi_get_cb_info(env, info, &argc, &argv, null, null) != c.napi_ok) {
-        translate.throw(env, "Failed to get args.") catch return null;
+    errdefer allocator.free(data);
+
+    // Parse the event objects from the array.
+    const events = std.mem.bytesAsSlice(Event, data[0 .. @sizeOf(Event) * array_length]);
+    for (events, 0..) |*event, i| {
+        const object = try translate.array_element(env, array, @intCast(i));
+        switch (Event) {
+            Account, Transfer => {
+                inline for (std.meta.fields(Event)) |field| {
+                    const FieldInt = switch (@typeInfo(field.type)) {
+                        .Struct => |info| info.backing_integer.?,
+                        else => field.type,
+                    };
+
+                    const value = try @field(translate, @typeName(FieldInt) ++ "_from_object")(
+                        env,
+                        object,
+                        @ptrCast(field.name ++ "\x00"),
+                    );
+
+                    if (std.mem.eql(u8, field.name, "timestamp") and value != 0) {
+                        return translate.throw(
+                            env,
+                            "Timestamp should be set to 0 as this will be set correctly by Replica.",
+                        );
+                    }
+
+                    @field(event, field.name) = switch (@typeInfo(field.type)) {
+                        .Struct => @as(field.type, @bitCast(value)),
+                        else => value,
+                    };
+                }
+            },
+            u128 => event.* = try translate.u128_from_value(env, object, "lookup"),
+            else => @compileError("invalid Event type"),
+        }
     }
 
-    if (argc != 1) translate.throw(
-        env,
-        "Function tick() requires 1 argument exactly.",
-    ) catch return null;
+    // Return only the memory for the Events. The total size of the bytes allocation can be inferred
+    // via `bytes[0..max(bytes.len, events.len * ResultFrom(Event))]`.
+    return std.mem.sliceAsBytes(events);
+}
 
-    const context_raw = translate.value_external(
-        env,
-        argv[0],
-        "Failed to get Client Context pointer.",
-    ) catch return null;
-    const context = contextCast(context_raw.?) catch return null;
+// For a given op, packet memory is allocated to hold
+// `max(sizeof(Event/Request) * n, sizeof(Result/Response))`. As long as the op and n are known, the
+// allocation size can be computed.
+//
+// When submitting the packet, the data_size is `sizeof(Event) * n`. On packet results, we overwrite
+// the memory with `sizeof(Result) * completed` where completed <= n. If less Results are completed
+// than Events, the completed size no longer reflects "n". So instead, data_size stores
+// `packed(n, completed)`: arg1 to know alloc size, arg2 to know what to parse back to JS.
+const DataSize = packed struct(u32) {
```

Comment:
Let's call it `BufferSize`, "Data" can be misleading since it holds events and results.

```suggestion
const BufferSize = packed struct(u32) {
```